packages/solid/store/src/store.ts

```diff
@@ -215,171 +215,150 @@ export function updatePath(current: StoreNode, path: any[], traversed: (keyof an
   } else setProperty(current, part, value);
 }
 
-export type DeepReadonly<T> = {
-  readonly [K in keyof T]: T[K] extends NotWrappable ? T[K] : DeepReadonly<T[K]>;
-};
-
-export type StoreSetter<T> =
-  | T
-  | Partial<T>
-  | ((prevState: T, traversed?: (keyof any)[]) => Partial<T> | void);
+export type DeepReadonly<T> = 0 extends 1 & T
+  ? T
+  : {
+      readonly [K in keyof T]: T[K] extends NotWrappable ? T[K] : DeepReadonly<T[K]>;
+    };
+export type DeepMutable<T> = 0 extends 1 & T
+  ? T
+  : {
+      -readonly [K in keyof T]: T[K] extends NotWrappable ? T[K] : DeepMutable<T[K]>;
+    };
 
 export type StorePathRange = { from?: number; to?: number; by?: number };
 
-export type ArrayFilterFn<T> = (item: T, index: number) => boolean;
+export type ArrayFilterFn<T> = (item: DeepReadonly<T>, index: number) => boolean;
 
-export type Part<T> = [T] extends [never]
-  ? never
-  : [keyof T] extends [never]
-  ? never
-  :
-      | keyof T
-      | (keyof T)[]
-      | (number extends keyof T ? ArrayFilterFn<T[number]> | StorePathRange : never);
+export type StoreSetter<T, U extends PropertyKey[] = []> =
+  | ((
+      prevState: DeepReadonly<T>,
+      traversed: U
+    ) => DeepReadonly<T> | Partial<DeepReadonly<T>> | void)
+  | DeepReadonly<T>
+  | Partial<DeepReadonly<T>>;
 
-export type Next<T, K extends Part<T>> = [K] extends [never]
-  ? never
-  : K extends keyof T
-  ? T[K]
-  : K extends (keyof T)[]
-  ? T[K[number]]
-  : // since K extends Part<T> and we have excluded never,
-  // we assume that K is now ArrayFilterFn or StorePathRange
-  number extends keyof T
-  ? T[number]
-  : never;
+export type Part<T, K extends KeyOf<T> = KeyOf<T>> = [K] extends [never]
+  ? never // return never if key is never, else it'll return readonly never[] as well
+  : K | readonly K[] | (number extends K ? ArrayFilterFn<T[number]> | StorePathRange : never);
 
-export type WrappableNext<T, K extends Part<T>> = Exclude<Next<T, K>, NotWrappable>;
+// shortcut to avoid writing `Exclude<T, NotWrappable>` too many times
+type W<T> = Exclude<T, NotWrappable>;
 
-type DistributeRest<T, K extends Part<T>> = K extends K ? [K, ...Rest<Next<T, K>>] : never;
-export type Rest<T> = 0 extends 1 & T
-  ? [...(keyof any)[], any]
-  : [StoreSetter<T>] | (T extends NotWrappable ? never : DistributeRest<T, Part<T>>);
+// specially handle keyof to avoid errors with arrays and any
+type KeyOf<T> = number extends keyof T
+  ? 0 extends 1 & T // if it's any just return keyof T
+    ? keyof T
+    : [T] extends [never] // keyof never is PropertyKey which number extends; return never
+    ? never
+    : number // it's not any or never so it's an array or tuple; exclude the non-number properties
+  : keyof T;
```

Comment:
This solves a lot of issues, as originally using `keyof T` would break with arrays, as it included other undesired array properties. By using a custom `KeyOf` instead, we can simplify most of the other types greatly. Not sure if this should be exported.